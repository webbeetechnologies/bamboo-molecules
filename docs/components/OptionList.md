# OptionList

Molecules exposes a highly performant memoized OptionList hook that can support large list of options. The hook returns a SectionList element.

[Video explaining OptionList](https://www.loom.com/share/42de0e5f4f814bd8ac8b851a01a633a7)

## Personas

-   End User
-   Component Consumer

### End User

-   As an end user, I expect that the items can be grouped.
-   As an end user, I want to be able to see large performant list of options.
-   As an end user, I want to see `"No options to display"` when there are no options to be displayed.

### Component Consumer

-   As a component consumer, I want to show a list of options or list of grouped options for user input.
-   As a component consumer, I want to be able to allow selection of multiple options with `multiple` boolean prop.
-   As a component consumer, I may not always want to pass the `renderOption` function and expect the component to use `item.label` for labels and `item.value` property for value in ListItem
-   As a component consumer, I expect OptionList to wrap a [SectionList](./SectionList.md).
-   As a component consumer, I expect the `renderItem` function extends the functionality of renderItem from SectionList and also adds an the `onClick` function to the arguments.

## Interface

```ts
import { Data, SectionData } from 'SectionList';
import type { UseSearchableProps } from 'useSearchable';
type OptionListProps = UseSearchableProps & UseDataSourceProps &
    SectionListProps & {
        onPress?: (item: Data) => void;
        renderItem?: (item: Data, index: number, onPress: () => void) => ReactNode;
    };
```

## Implementation Details

-   Returns a SectionList that can be used to create any of the following type of components.
    -   [Checkbox Group](./Checkbox.md#checkbox-group)
    -   [Radio Group](./Radio.md#radio-group)
    -   [SelectInput](./SelectInput.md)
    -   Maybe also [Chips List](./Chips.md#chips-list)
