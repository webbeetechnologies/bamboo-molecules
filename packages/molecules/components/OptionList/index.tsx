import { getRegisteredComponentWithFallback, registerMoleculesComponents } from '../../core';
import OptionListDefault from './OptionList';

registerMoleculesComponents({
    OptionList: OptionListDefault,
});

export const OptionList = getRegisteredComponentWithFallback('OptionList', OptionListDefault);

export { Props as OptionListProps, IOptionList, OptionListRenderItemInfo } from './OptionList';
export { optionListStyles } from './utils';
