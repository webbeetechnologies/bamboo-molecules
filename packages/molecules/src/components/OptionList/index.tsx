export {
    default as OptionList,
    Props as OptionListProps,
    IOptionList,
    OptionListRenderItemInfo,
} from './OptionList';
export { optionListStyles } from './utils';
