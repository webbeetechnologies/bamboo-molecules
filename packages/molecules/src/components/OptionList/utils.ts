import { getRegisteredMoleculesComponentStyles, registerComponentsStyles } from '../../core';
import { StyleSheet } from 'react-native-unistyles';

const optionListStylesDefault = StyleSheet.create(theme => ({
    container: {
        flex: 1,
    },
    searchInputContainer: {
        backgroundColor: theme.colors.surface,
    },
}));

registerComponentsStyles({
    optionListStyles: optionListStylesDefault,
});

export const optionListStyles = getRegisteredMoleculesComponentStyles('optionListStyles');
