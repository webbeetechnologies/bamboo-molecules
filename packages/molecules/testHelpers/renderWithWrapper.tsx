import type { ReactElement, JSXElementConstructor, ReactNode } from 'react';
import { render, RenderOptions } from '@testing-library/react-native';
import { ProvideMolecules } from '../src';

export const renderWithWrapper = <P extends {} = {}>(
    component: ReactElement<P, string | JSXElementConstructor<P>>,
    options?: RenderOptions,
): ReturnType<typeof render> => {
    const Wrapper = ({ children }: { children: ReactNode }) => (
        <ProvideMolecules>{children}</ProvideMolecules>
    );

    return render(component, { wrapper: Wrapper, ...options });
};
