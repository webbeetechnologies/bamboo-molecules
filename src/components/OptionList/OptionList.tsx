import { memo, PropsWithoutRef, ReactElement, RefAttributes, useCallback, useMemo } from 'react';
import type { ViewStyle } from 'react-native';
import type { FlashList } from '@shopify/flash-list';
import {
    useComponentStyles,
    useControlledValue,
    useMolecules,
    useSearchable,
    UseSearchableProps,
} from '../../hooks';
import type { SectionListProps } from '../SectionList';
import type { SectionListRenderItemInfo } from '../SectionList/types';

type DefaultSectionT<TItem> = {
    data?: TItem[];
    [key: string]: any;
};

type DefaultItemT = {
    [key: string]: any;
};

// To make a correct type inference
export type IOptionList = <
    ItemType extends DefaultItemT = DefaultItemT,
    TSectionType extends DefaultSectionT<ItemType> = DefaultSectionT<ItemType>,
>(
    props: PropsWithoutRef<Props<ItemType, TSectionType>> & RefAttributes<FlashList<ItemType>>,
) => ReactElement;

export type Props<TItem, TSection> = UseSearchableProps &
    Omit<SectionListProps<TItem, TSection>, 'sections'> & {
        records: TSection[];
        containerStyle?: ViewStyle;
        searchInputContainerStyle?: ViewStyle;
        multiple?: boolean;
        selectable?: boolean;
        selectedItem?: TItem | TItem[];
        onSelectItemChange?: (item: TItem | TItem[]) => void;
    };

const OptionList = <
    TItem extends DefaultItemT = DefaultItemT,
    TSection extends DefaultSectionT<TItem> = DefaultSectionT<TItem>,
>({
    query = '',
    onQueryChange,
    searchInputProps,
    searchable,
    containerStyle = {},
    searchInputContainerStyle = {},
    style: styleProp,
    records,
    multiple = false,
    selectable,
    selectedItem: selectedItemProp,
    onSelectItemChange: onSelectItemChangeProp,
    renderItem: renderItemProp,
    ...rest
}: Props<TItem, TSection>) => {
    const { SectionList, View, TouchableRipple } = useMolecules();
    const SearchField = useSearchable({ query, onQueryChange, searchable, searchInputProps });
    const [selectedItem, onSelectItemChange] = useControlledValue<TItem | TItem[]>({
        value: selectedItemProp,
        onChange: onSelectItemChangeProp,
    });

    const componentStyles = useComponentStyles('OptionList', [
        { container: containerStyle, searchInputContainer: searchInputContainerStyle },
    ]);

    const { containerStyles, searchInputContainerStyles, style } = useMemo(() => {
        const { container, searchInputContainer, ...restStyle } = componentStyles;

        return {
            containerStyles: container,
            searchInputContainerStyles: searchInputContainer,
            style: [restStyle, styleProp],
        };
    }, [componentStyles, styleProp]);

    const onPressItem = useCallback(
        (item: TItem) => {
            onSelectItemChange(
                multiple ? [...(Array.isArray(selectedItem) ? selectedItem : []), item] : item,
            );
        },
        [multiple, onSelectItemChange, selectedItem],
    );

    const renderItem = useCallback(
        (info: SectionListRenderItemInfo<TItem, TSection>) => {
            return selectable && info.item?.selectable !== false ? (
                <TouchableRipple onPress={() => onPressItem(info.item)}>
                    {renderItemProp(info)}
                </TouchableRipple>
            ) : (
                renderItemProp(info)
            );
        },
        [TouchableRipple, onPressItem, renderItemProp, selectable],
    );

    return (
        <View style={containerStyles}>
            <>{SearchField && <View style={searchInputContainerStyles}>{SearchField}</View>}</>
            <SectionList {...rest} sections={records} renderItem={renderItem} style={style} />
        </View>
    );
};

export default memo(OptionList) as IOptionList;
