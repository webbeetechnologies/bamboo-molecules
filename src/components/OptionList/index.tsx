export { default as OptionList, Props as OptionListProps, IOptionList } from './OptionList';
export { optionListStyles } from './utils';
