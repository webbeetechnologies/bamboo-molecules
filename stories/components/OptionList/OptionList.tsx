import { useMolecules } from '../../../src';
import { OptionList, ListItem, OptionListProps } from '../../../src/components';
import { useCallback } from 'react';
import { StyleSheet } from 'react-native';

export type Props = OptionListProps & {};

export const Example = (props: Props) => {
    const { Text } = useMolecules();

    const renderItem = useCallback(
        ({ item }: any) => (
            <ListItem>
                <ListItem.Title>{item.title}</ListItem.Title>
            </ListItem>
        ),
        [],
    );
    const renderSectionHeader = useCallback(
        ({ section }: any) => <Text style={styles.sectionTitle}>{section.title}</Text>,
        [Text],
    );

    return (
        <OptionList {...props} renderItem={renderItem} renderSectionHeader={renderSectionHeader} />
    );
};

const styles = StyleSheet.create({
    sectionTitle: {
        fontSize: 'typescale.titleSmall.fontSize' as unknown as number,
        fontWeight: '600',
        paddingLeft: 'spacings.3',
        backgroundColor: 'colors.surface',
        paddingTop: 'spacings.2',
    },
});
