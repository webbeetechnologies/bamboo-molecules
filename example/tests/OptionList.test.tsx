import { Text } from 'react-native';
import { Example as OptionList } from '../stories/components/OptionList/OptionList';
import { renderWithWrapper } from '../testHelpers';

const renderItemMock = jest.fn(({ item }) => {
    return <Text>{item.text}</Text>;
});
const onQueryChangeMock = jest.fn();
const onSelectItemChangeMock = jest.fn();
const records = [
    {
        title: 'Numbers',
        data: [
            {
                text: '1',
            },
            {
                text: '2',
            },
            {
                text: '3',
            },
            {
                text: '4',
            },
        ],
    },
    {
        title: 'Letters',
        data: [
            {
                text: 'A',
            },
            {
                text: 'B',
            },
            {
                text: 'C',
            },
            {
                text: 'D',
            },
        ],
    },
];

describe('OptionList', () => {
    it('renders default OptionList with records', () => {
        const tree = renderWithWrapper(
            <OptionList records={records} renderItem={renderItemMock} />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('renders OptionList with SearchField', () => {
        const tree = renderWithWrapper(
            <OptionList
                searchable
                onQueryChange={onQueryChangeMock}
                records={records}
                renderItem={renderItemMock}
            />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('renders OptionList with selectable items', () => {
        const tree = renderWithWrapper(
            <OptionList
                records={records}
                renderItem={renderItemMock}
                selectable
                onSelectItemChange={onSelectItemChangeMock}
            />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
