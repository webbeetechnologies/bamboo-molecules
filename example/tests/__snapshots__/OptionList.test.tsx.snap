// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionList renders OptionList with SearchField 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 8,
      }
    }
  >
    <View
      style={Object {}}
    >
      <View
        style={
          Object {
            "animationScale": 1,
            "backgroundColor": "rgba(231, 224, 236, 1)",
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
            "flexDirection": "row",
            "paddingHorizontal": 16,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "rgba(73, 69, 79, 1)",
              "bottom": 0,
              "height": 1,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "zIndex": 1,
            }
          }
          testID="text-input-underline"
        />
        <View
          style={
            Object {
              "flexGrow": 1,
              "flexShrink": 1,
              "minHeight": 56,
              "paddingBottom": 0,
              "paddingTop": 0,
            }
          }
        >
          <TextInput
            editable={true}
            maxFontSizeMultiplier={15}
            multiline={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholderTextColor="rgba(73, 69, 79, 1)"
            size="md"
            style={
              Array [
                Object {
                  "color": "rgba(28, 27, 31, 1)",
                  "flexGrow": 1,
                  "fontSize": 16,
                  "fontWeight": "400",
                  "lineHeight": 24,
                  "paddingBottom": 8,
                  "paddingTop": 24,
                },
                Object {
                  "height": 56,
                },
                false,
                Object {
                  "textAlign": "left",
                  "textAlignVertical": "center",
                },
                false,
                undefined,
              ]
            }
            testID="text-input-flat"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Object {
              "color": "rgba(73, 69, 79, 1)",
              "iconSize": 24,
              "justifyContent": "center",
              "marginLeft": 12,
              "marginRight": -4,
            }
          }
          testID="text-input-right-element"
        >
          <View
            collapsable={false}
            style={
              Object {
                "alignSelf": undefined,
                "bottom": undefined,
                "left": undefined,
                "position": undefined,
                "right": undefined,
                "shadowColor": "#000",
                "shadowOffset": Object {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
                "top": undefined,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "shadowColor": "#000",
                  "shadowOffset": Object {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0,
                  "shadowRadius": 0,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  Object {
                    "backgroundColor": undefined,
                    "borderColor": "rgba(121, 116, 126, 1)",
                    "borderRadius": 750,
                    "borderWidth": undefined,
                    "height": 34,
                    "margin": undefined,
                    "overflow": "hidden",
                    "width": 34,
                  }
                }
              >
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    Object {
                      "disabled": false,
                    }
                  }
                  accessible={true}
                  focusable={true}
                  hitSlop={
                    Object {
                      "bottom": 6,
                      "left": 6,
                      "right": 6,
                      "top": 6,
                    }
                  }
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Array [
                      Object {
                        "overflow": "hidden",
                      },
                      Object {
                        "alignItems": "center",
                        "flexGrow": 1,
                        "justifyContent": "center",
                      },
                    ]
                  }
                >
                  <Text
                    color="rgba(73, 69, 79, 1)"
                    name="magnify"
                    size={24}
                    style={
                      Object {
                        "animationScale": 1,
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "data": Array [
            Object {
              "text": "1",
            },
            Object {
              "text": "2",
            },
            Object {
              "text": "3",
            },
            Object {
              "text": "4",
            },
          ],
          "title": "Numbers",
        },
        Object {
          "data": Array [
            Object {
              "text": "A",
            },
            Object {
              "text": "B",
            },
            Object {
              "text": "C",
            },
            Object {
              "text": "D",
            },
          ],
          "title": "Letters",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={
      Array [
        0,
        6,
      ]
    }
    style={Object {}}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      />
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          1
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          2
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          3
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          4
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      />
      <View
        onLayout={[Function]}
        style={null}
      />
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          A
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          B
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          C
        </Text>
      </View>
      <View
        style={
          Object {
            "height": 0,
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;

exports[`OptionList renders OptionList with selectable items 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "data": Array [
            Object {
              "text": "1",
            },
            Object {
              "text": "2",
            },
            Object {
              "text": "3",
            },
            Object {
              "text": "4",
            },
          ],
          "title": "Numbers",
        },
        Object {
          "data": Array [
            Object {
              "text": "A",
            },
            Object {
              "text": "B",
            },
            Object {
              "text": "C",
            },
            Object {
              "text": "D",
            },
          ],
          "title": "Letters",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onSelectItemChange={[MockFunction]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={
      Array [
        0,
        6,
      ]
    }
    style={Object {}}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      />
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              false,
              Object {},
            ]
          }
        >
          <Text>
            1
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              false,
              Object {},
            ]
          }
        >
          <Text>
            2
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              false,
              Object {},
            ]
          }
        >
          <Text>
            3
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              false,
              Object {},
            ]
          }
        >
          <Text>
            4
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      />
      <View
        onLayout={[Function]}
        style={null}
      />
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              false,
              Object {},
            ]
          }
        >
          <Text>
            A
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              false,
              Object {},
            ]
          }
        >
          <Text>
            B
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              false,
              Object {},
            ]
          }
        >
          <Text>
            C
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "height": 0,
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;

exports[`OptionList renders default OptionList with records 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "data": Array [
            Object {
              "text": "1",
            },
            Object {
              "text": "2",
            },
            Object {
              "text": "3",
            },
            Object {
              "text": "4",
            },
          ],
          "title": "Numbers",
        },
        Object {
          "data": Array [
            Object {
              "text": "A",
            },
            Object {
              "text": "B",
            },
            Object {
              "text": "C",
            },
            Object {
              "text": "D",
            },
          ],
          "title": "Letters",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={
      Array [
        0,
        6,
      ]
    }
    style={Object {}}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      />
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          1
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          2
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          3
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          4
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      />
      <View
        onLayout={[Function]}
        style={null}
      />
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          A
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          B
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text>
          C
        </Text>
      </View>
      <View
        style={
          Object {
            "height": 0,
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;
